op-conductor/health/monitor.go

```diff
@@ -0,0 +1,138 @@
+package health
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/ethereum-optimism/optimism/op-node/p2p"
+	"github.com/ethereum-optimism/optimism/op-node/rollup"
+	"github.com/ethereum-optimism/optimism/op-service/dial"
+)
+
+// HealthMonitor defines the interface for monitoring the health of the sequencer.
+type HealthMonitor interface {
+	// Subscribe returns a channel that will be notified for every health check.
+	Subscribe() <-chan bool
+	// Start starts the health check.
+	Start() error
+	// Stop stops the health check.
+	Stop() error
+}
+
+// NewSequencerHealthMonitor creates a new sequencer health monitor.
+// interval is the interval between health checks measured in seconds.
+// safeInterval is the interval between safe head progress measured in seconds.
+// minPeerCount is the minimum number of peers required for the sequencer to be healthy.
+func NewSequencerHealthMonitor(log log.Logger, interval, safeInterval, minPeerCount uint64, rollupCfg *rollup.Config, node dial.RollupClientInterface, p2p p2p.API) HealthMonitor {
+	return &SequencerHealthMonitor{
+		log:            log,
+		done:           make(chan struct{}),
+		interval:       interval,
+		healthUpdateCh: make(chan bool),
+		rollupCfg:      rollupCfg,
+		safeInterval:   safeInterval,
+		minPeerCount:   minPeerCount,
+		node:           node,
+		p2p:            p2p,
+	}
+}
+
+// SequencerHealthMonitor monitors sequencer health.
+type SequencerHealthMonitor struct {
+	log  log.Logger
+	done chan struct{}
+	wg   sync.WaitGroup
+
+	rollupCfg      *rollup.Config
+	safeInterval   uint64
+	minPeerCount   uint64
+	interval       uint64
+	healthUpdateCh chan bool
+
+	node dial.RollupClientInterface
+	p2p  p2p.API
+}
+
+var _ HealthMonitor = (*SequencerHealthMonitor)(nil)
+
+// Start implements HealthMonitor.
+func (hm *SequencerHealthMonitor) Start() error {
+	hm.log.Info("starting health monitor")
+	hm.wg.Add(1)
+	go hm.loop()
+
+	hm.log.Info("health monitor started")
+	return nil
+}
+
+// Stop implements HealthMonitor.
+func (hm *SequencerHealthMonitor) Stop() error {
+	hm.log.Info("stopping health monitor")
+	close(hm.done)
+	hm.wg.Wait()
+
+	hm.log.Info("health monitor stopped")
+	return nil
+}
+
+// Subscribe implements HealthMonitor.
+func (hm *SequencerHealthMonitor) Subscribe() <-chan bool {
+	return hm.healthUpdateCh
+}
+
+func (hm *SequencerHealthMonitor) loop() {
+	defer hm.wg.Done()
+
+	duration := time.Duration(hm.interval) * time.Second
+	ticker := time.NewTicker(duration)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-hm.done:
+			return
+		case <-ticker.C:
+			hm.healthUpdateCh <- hm.healthCheck()
+		}
+	}
+}
+
+// healthCheck checks the health of the sequencer by 3 criteria:
+// 1. unsafe head is progressing per block time
+// 2. safe head is progressing every configured batch submission interval
+// 3. peer count is above the configured minimum
+func (hm *SequencerHealthMonitor) healthCheck() bool {
```

Comment:
I notice this pings the `op-node` RPC API, would it also be useful to ping the EL RPC API? This would ensure that both services are actually running

Reply:
We definitely could. 
My current understanding is that unsafeHead will not proceed if EL is not healthy, can somebody help confirm? If not, we can add an extra step here to ping EL RPC to make sure it is alive.

Reply:
Correct, unsafehead cannot proceed unless the EL is healthy.